### PR TITLE
test: Upgrade to ginkgo v2

### PR DESCRIPTION
**Pull request by qinqon**
**What this PR does / why we need it**:

Migrate to ginkgo v2 https://onsi.github.io/ginkgo/MIGRATING_TO_V2#upgrading-to-ginkgo-20.

- Remove table extension package
- Adapt reporter to specResult
- Remove polarion reporter (this is no longer par of tier1)
- Change imports

**Special notes for your reviewer**:

**Release note**:

```release-note
Migrate to ginkgo v2
```

### CODE REVIEW

File: api/go.mod
Comment (oshoval) [inline api/go.mod:6]: maybe worth to add replace block at the end, so it will pin the version?
same on the ./go.mod

File: Makefile
Comment (oshoval) [inline Makefile:34]: The new report has little different XSD, as long as you are fine with it, it is ok
you might want to specify the --output-dir $ARTIFACTS ? or the current directory is fine ?

File: Makefile
Comment (oshoval) [inline Makefile:195]: maybe worth to have ENV_VAR for the 1.17 here and in the next line ?

File: automation/check-patch.e2e-operator-k8s.sh
Comment (oshoval) [inline automation/check-patch.e2e-operator-k8s.sh:31]: from what i saw i had to supply $ARTIFACTS as --output-dir, not as part of --junit-report
unless it works for you

File: test/e2e/handler/nnce_conditions_test.go
Comment (oshoval) [inline test/e2e/handler/nnce_conditions_test.go:40]: I saw that the new specReport now starts with Texts[0] instead of Texts[1]
for the old reporters 
https://github.com/nmstate/kubernetes-nmstate/pull/1002/commits/3a5358141b37477b132f2751393339653827e0c9#diff-6b11a0d0638c5aea0a09de9f7fa737dbfee413cf63d52940c706aa7c4b486dbdR136

worth to check that the whole string you expect appears please
(start with `EnactmentCondition`)

i don't refer to the removed `rfe_id`, just to what the junit starts with now

File: test/e2e/operator/main_test.go
Comment (oshoval) [inline test/e2e/operator/main_test.go:28]: can be dot import or gives a conflict ?
or maybe even nameless `"github.com/onsi/ginkgo/v2/types"`

File: test/reporter/reporter.go
Comment (oshoval) [inline test/reporter/reporter.go:62]: maybe should start with 0 ?
i saw that else it omit text

**Review by oshoval — COMMENTED** [review]
Thanks

File: api/go.mod
Comment (qinqon) [reply api/go.mod:6]: The `sed` command did the job `replace` not needed.

Comment (oshoval) [reply api/go.mod:6]: yes indeed, it is just for next time, so no surprise change happens, but whatever u prefer

Comment (qinqon) [reply api/go.mod:6]: Prefer having it explicit is very misleading to see an "import" that does not have the stuff from the source file.

File: Makefile
Comment (qinqon) [reply Makefile:34]: Let's see how prow consume it, the output-dir is fine I think, let's check with CI, is the only real user.

Comment (qinqon) [reply Makefile:34]: Let's wait for CI and see how it goes.

File: Makefile
Comment (qinqon) [reply Makefile:195]: Reading it from go.mod

File: automation/check-patch.e2e-operator-k8s.sh
Comment (qinqon) [reply automation/check-patch.e2e-operator-k8s.sh]: I think it works here.

File: Makefile
Comment (oshoval) [reply Makefile:195]: so it means that once go.mod is changed this need to be changed as well
maybe at least a comment on go.mod then ?

File: test/e2e/operator/main_test.go
Comment (qinqon) [reply test/e2e/operator/main_test.go:28]: We cannot import types with dot
```
test/e2e/handler/main_test.go:28:2: `Report` redeclared in this block (typecheck)
	. "github.com/onsi/ginkgo/v2/types"
```

File: test/reporter/reporter.go
Comment (oshoval) [inline test/reporter/reporter.go:62]: simply `specReport.ContainerHierarchyTexts`

File: Makefile
Comment (oshoval) [inline Makefile:4]: double space after `=`

**Comment by oshoval** [comment]
hack/go-version.sh is missing or it is in this PR ?

**Comment by qinqon** [comment]
> hack/go-version.sh is missing or it is in this PR ? if it is missing, it means there isn't a vendor CI job

Adding it.

File: Makefile
Comment (oshoval) [inline Makefile:43]: 60s isnt it ?

**Comment by oshoval** [comment]
Junit missing here
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/1002/pull-kubernetes-nmstate-e2e-operator-k8s/1501516752465432576

File: Makefile
Comment (qinqon) [reply Makefile:43]: The --help show a pair of them:

```bash
[ellorent@fedora kubernetes-nmstate]$ go run github.com/onsi/ginkgo/v2/ginkgo run --help |grep thres
  --slow-spec-threshold [duration] (default: 5s)
```

**Comment by qinqon** [comment]
> Junit missing here (might be because of output folder, and if it is there, it is because junit-report has different XSD scheme than what prow SpyGlass looks for) https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/1002/pull-kubernetes-nmstate-e2e-operator-k8s/1501516752465432576

Junit report is there https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/1002/pull-kubernetes-nmstate-unit-test/1501516751844675584/artifacts/unit-test-junit.xml, we may need to adjust prow job.

**Comment by oshoval** [comment]
> > Junit missing here (might be because of output folder, and if it is there, it is because junit-report has different XSD scheme than what prow SpyGlass looks for) https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/1002/pull-kubernetes-nmstate-e2e-operator-k8s/1501516752465432576
> 
> Junit report is there https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/1002/pull-kubernetes-nmstate-unit-test/1501516751844675584/artifacts/unit-test-junit.xml, we may need to adjust prow job.

This is why i think we should first use the old reporter, and on a follow PR adapt spyglass and the new reporter.
if you run xmllint you will see the new report has fields that prow doesn't expect, hence fails to parse

File: automation/check-patch.unit-test.sh
Comment (oshoval) [inline automation/check-patch.unit-test.sh:11]: the --output-dir can be used instead this copy ?

Comment (qinqon) [reply automation/check-patch.unit-test.sh:11]: Ahh right!

File: Makefile
Comment (oshoval) [reply Makefile]: both works it seems (or at least doesn't panic), maybe worth to be consistent with the other params in the same line ?

File: automation/check-patch.unit-test.sh
Comment (qinqon) [reply automation/check-patch.unit-test.sh:11]: Ahh no if we do so they will be overwitten the root and api.

Comment (oshoval) [reply automation/check-patch.unit-test.sh:11]: can't each have other file name ?
unless i am missing something, not sure about the flow here

Comment (qinqon) [reply automation/check-patch.unit-test.sh]: This is what I did

Comment (qinqon) [reply automation/check-patch.unit-test.sh]: Spyglass expect files like the following
```yaml
- lens:                                                                     
        name: junit                                                             
      required_files:                                                           
        - artifacts/junit.*\.xml         
```

Comment (oshoval) [reply automation/check-patch.unit-test.sh]: so just rename it right ?
but i think that even if you will have the right file, it will fail parsing it because the XSD is changed

**Comment by oshoval** [comment]
can you please inject an error and see the expected output appears on prow junit ?
asking because i saw the error field wasn't printed the same with the new `--junit-report` of ginkgo 2.0

**Comment by qinqon** [comment]
> can you please inject an error and see the expected output appears on prow junit spyglass ? asking because i saw the error field wasn't printed good with the new `--junit-report` of ginkgo 2.0 (the XML entity of it didn't have the same structure)

Sure, the possitive case works fine I see spyglass understanding the junit format.

**Comment by oshoval** [comment]
> Sure, the possitive case works fine I see spyglass understanding the junit format.

Do you assume there won't be skip tests in the future ? 
if so we are fine in that aspect, else might good to check if when there is a skip it still reads it

**Comment by qinqon** [comment]
We need this to have "message" attribute from json xml report to be presented https://github.com/GoogleCloudPlatform/testgrid/pull/838

**Comment by qinqon** [comment]
@oshoval @rhrazdil CI is fine now, can you re-review ?

**Comment by rhrazdil** [comment]
The code and the test output looks good 
/lgtm

**Comment by qinqon** [comment]
/hold
Using ginkgo v2 increasing handler e2e tests CI job to 1 hour more
from 1h30m18s to 2h27m37s

**Comment by qinqon** [comment]
/hold cancel
Job was waitting before start

**Comment by qinqon** [comment]
@oshoval can you review ?

**Comment by oshoval** [comment]
Don't we want to wait until this is merged + image bumped ?
https://github.com/nmstate/kubernetes-nmstate/pull/1002#issuecomment-1063904759

else it will be less comfortable to see about what test failed
(will continue review in a bit)

File: nncp.yaml
Comment (oshoval) [inline nncp.yaml:1]: this file doesnt belong right ?

File: Makefile
Comment (oshoval) [inline Makefile:35]: about `--randomize-all --randomize-suite`
aint better to have a deterministic run order? (as we are doing on kubevirt by fixing the seed)
(can be on a follow PR if desired)

**Review by oshoval — COMMENTED** [review]
2nd commit has vendor folder and go.mod / go.sum and such that can be squashed to the first if you want (else it is little confusing, as some modules were updated on the first and removed on the 2nd commit, i.e kubevirt/qe-tools but whatever you prefer)
best will be that all non vendor files are in first commit, and all vendor files are in 2nd
this way we won't miss files / changes.

https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/1002/pull-kubernetes-nmstate-e2e-handler-k8s-future/1502277570844954624
This one sometimes doesn't finish in 3 hours, can you please check if it is due to this PR ?

**Comment by oshoval** [comment]
it failed again (takes more than 3h and gets terminated)
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/1002/pull-kubernetes-nmstate-e2e-handler-k8s-future/1503002210584236032

due to this PR or not ?

**Comment by qinqon** [comment]
@oshoval looks like gingov2 has increase ci by one hour, does it sound familiar to you? At your kubevirt PR you are incrementing time too

**Comment by oshoval** [comment]
> @oshoval looks like gingov2 has increase ci by one hour, does it sound familiar to you? At your kubevirt PR you are incrementing time too

It doesn't even starts the tests here does it ?
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/1002/pull-kubernetes-nmstate-e2e-handler-k8s-future/1503002210584236032

On kubevirt only a few were incremented by few minutes (for example because deprecated features like async running), most stayed the same, i don't think it is the same problem we see here

**Comment by qinqon** [comment]
> > @oshoval looks like gingov2 has increase ci by one hour, does it sound familiar to you? At your kubevirt PR you are incrementing time too
> 
> It doesn't even starts the tests here does it ? https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/1002/pull-kubernetes-nmstate-e2e-handler-k8s-future/1503002210584236032
> 
> On kubevirt only a few were incremented by few minutes (for example because deprecated features like async running), most stayed the same, i don't think it is the same problem we see here

Right, it got stuck at ARM build, sometimes it takes too much to for the ARM emulation with qemu, anyhow, can you review ? looks like its not related.

**Comment by oshoval** [comment]
> > > @oshoval looks like gingov2 has increase ci by one hour, does it sound familiar to you? At your kubevirt PR you are incrementing time too
> > 
> > 
> > It doesn't even starts the tests here does it ? https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/1002/pull-kubernetes-nmstate-e2e-handler-k8s-future/1503002210584236032
> > On kubevirt only a few were incremented by few minutes (for example because deprecated features like async running), most stayed the same, i don't think it is the same problem we see here
> 
> Right, it got stuck at ARM build, depending on the prow node it takes longer at QEMU emulating and compiling for ARM, anyhow, can you review ? looks like its not related.

I already reviewed yesterday
https://github.com/nmstate/kubernetes-nmstate/pull/1002#pullrequestreview-908115318
https://github.com/nmstate/kubernetes-nmstate/pull/1002#issuecomment-1066041719

File: nncp.yaml
Comment (qinqon) [reply nncp.yaml:1]: Removing

File: Makefile
Comment (oshoval) [reply Makefile:35]: ye i understand, just on kubevirt (at least e2e) we are doing it ordered,
anyhow we have less flakes and it is unit tests so it is less problematic + whatever you prefer

**Comment by qinqon** [comment]
> Don't we want to wait until this is merged + image bumped ? [#1002 (comment)](https://github.com/nmstate/kubernetes-nmstate/pull/1002#issuecomment-1063904759)
> 
> else it will be less comfortable to see about what test failed (will continue review in a bit)

This is massive already let's merge it so people start rebasing PRs, while the junit parsing at prow get improved, we can open an issue to track this.

**Review by oshoval — APPROVED** [review]
Thanks

maybe worth to already open an issue about the errors ?
it can explain that in order to see the interesting error text the log can be expanded meanwhile

File: Makefile
Comment (qinqon) [reply Makefile:35]: e2e is another kind of anymal.

**Review by rhrazdil — APPROVED** [review]

**Comment by oshoval** [comment]
those two failures happens a lot, maybe worth to understand if it also on main or not ?

**Comment by rhrazdil** [comment]
Happens on main too

**Pull request merged**
